HTTP+JSON serialization sink

theatrus opened the pull request:
In order to reduce pull request size and interdependence,
here is an incremental step for a stdout JSON output sink. The stdout
writes will of course be replaced with a lifoq push for libcurl posting.

The current sink infrastructure is a bit hard to unit test - once the
HTTP plumbing is in place, there is an integration test for it.

Commit:
HTTP+JSON serialization sink

In order to reduce pull request size and interdependence,
here is an incremental step for a stdout JSON output sink. The stdout
writes will of course be replaced with a lifoq push for libcurl posting.

The current sink infrastructure is a bit hard to unit test - once the
HTTP plumbing is in place, there is an integration test for it.

Commit:
Reduce boilerplate in serialization with strbuf

Commit:
Form serialization

smparkes commented on src/sink.c:
```diff
@@ -3,16 +3,26 @@
+        case SINK_TYPE_HTTP:
+        {
+            sink* actual_sink = init_http_sink((sink_config_http*)sc, config);
+            actual_sink->next = *sinks;
```
Could factor out the list plumbing.

smparkes replied:
factor list plumbing?

theatrus replied:
Yeah, I'll look at some macros for single LL intrusive operations. Its mostly straightforward. I'll open an issue for this one as a number of places could benefit.

smparkes commented on src/sink_http.c:
```diff
@@ -0,0 +1,200 @@
+
+/*
+ * Data from the metrics_iter callback */
+struct cb_info {
```
typedef for consistency?

theatrus replied:
Trying to leave typedefs for exported types, so I removed the top typedef.

smparkes commented on src/sink_http.c:
```diff
@@ -0,0 +1,200 @@
+        suffixed[base_len - 1] = '\0';                                  \
+        strcat(suffixed, suf);                                          \
+        json_object_set_new(obj, suffixed, val);                        \
+    } while(0);
```
No ;?

smparkes commented on src/sink_http.c:
```diff
@@ -0,0 +1,200 @@
+    case COUNTER:
+    {
+        if (config->extended_counters) {
+            char suffixed[base_len + 8];
```
factor our magic 8 and 20 below to constants with comments on the max value (esp. for timers below where it's more complicated)?

theatrus replied:
Now converted to snprintfs with explicit terminators

smparkes commented on src/sink_http.c:
```diff
@@ -0,0 +1,200 @@
+    return 0;
+}
+
+static int serialize_metrics(http_sink* sink, metrics* m, void* data) {
```
Be nice to have tests for this.

theatrus replied:
the whole serialize workflow? Yeah...

A bit awkward with how the test suite is run right now, where testing is only possible on public interfaces. 

I'm going to add a second constructor interface so I don't leak all of the internal state to a test and check serialization output.

Commit:
Cleanups and minor fixes

Commit:
HTTP support inlined and rebased

Commit:
Upgrades invoke non-existent configtest

Commit:
Allow specifying ciphers in config

Commit:
Init and teardown libcurl globally

Commit:
Add user agent data and a close header to requests

Commit:
More logging about how curl is setup

Commit:
Use a custom version of CURL when building RPM

Currently, on Red Hat derived systems, libcurl+NSS leaks memory. By
using a re-packaged version of libcurl which is build with OpenSSL, this
leak is avoided.

For details, see:

https://bugzilla.redhat.com/show_bug.cgi?id=1057388
